flow/layers/cacheable_layer.h

```diff
@@ -0,0 +1,106 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FLOW_LAYERS_CACHEABLE_LAYER_H_
+#define FLUTTER_FLOW_LAYERS_CACHEABLE_LAYER_H_
+
+#include <memory>
+#include "flutter/flow/embedded_views.h"
+#include "flutter/flow/layers/container_layer.h"
+#include "flutter/flow/raster_cache.h"
+#include "flutter/flow/raster_cacheable_entry.h"
+#include "include/core/SkColor.h"
+#include "include/core/SkMatrix.h"
+#include "include/core/SkPoint.h"
+#include "include/core/SkRect.h"
+
+namespace flutter {
+
+class Cacheable {
+ public:
+  Cacheable() = default;
+
+  void InitialCacheableLayerItem(Layer*, int);
+
+  void InitialCacheableDisplayListItem(DisplayList*,
+                                       const SkPoint& offset,
+                                       bool is_complex,
+                                       bool will_change);
+
+  void InitialCacheableSkPictureItem(SkPicture*,
+                                     const SkPoint& offset,
+                                     bool is_complex,
+                                     bool will_change);
+
+  class AutoCache {
+   public:
+    static AutoCache Create(Cacheable* cacheable,
+                            PrerollContext* context,
+                            const SkMatrix& matrix);
+
+    static AutoCache Create(DisplayListLayer* display_list,
+                            PrerollContext* context,
+                            const SkMatrix& matrix);
+
+    static AutoCache Create(PictureLayer* display_list,
+                            PrerollContext* context,
+                            const SkMatrix& matrix);
+
+    ~AutoCache();
+
+   private:
+    AutoCache(Cacheable* cacheable,
+              PrerollContext* context,
+              const SkMatrix& matrix)
+        : cacheable_(cacheable), context_(context), matrix_(matrix) {
+      if (context_ && context_->raster_cache) {
+        current_index_ = context_->raster_cached_entries->size();
+      }
+    }
+
+    int current_index_;
+    Cacheable* cacheable_ = nullptr;
+    PrerollContext* context_ = nullptr;
+    const SkMatrix& matrix_;
+  };
+
+  virtual Layer* asLayer() = 0;
+
+  virtual ~Cacheable() = default;
+
+  LayerCacheableItem* GetCacheableLayer() {
```

Comment:
This is not a good use of asFoo() and asBar() types of methods. The various layers called a specific method to instantiate a specific sub-type of the CacheableItem and then this class hides that type-specific knowledge and makes them ask permission to access the type of object that they specifically called into being.

Cacheable has gone from an unnecessary dispatch class to actively thwarting the layers from directly using their own style of cache item.

Cacheable needs to be deleted. Just have the layers have a "FooCacheableItem" field and instantiate it directly in their constructors, and then they'll never have to call these methods which use a de-obfuscation facility to perform the work we know should happen.